File: android/src/main/kotlin/dev/gece/imaplayer/ImaPlayerViewFactory.kt
package dev.gece.imaplayer

import android.content.Context
import android.net.Uri
import android.os.Build
import android.util.Log
import androidx.annotation.RequiresApi
import com.google.ads.interactivemedia.v3.api.ImaSdkFactory
import io.flutter.FlutterInjector
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import java.io.File

class ImaPlayerViewFactory(private val messenger: BinaryMessenger) :
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    @RequiresApi(Build.VERSION_CODES.N)
    override fun create(context: Context, id: Int, args: Any?): PlatformView {
        val payload = args as Map<String, Any>;

        val adsLoaderSettings = payload["ads_loader_settings"] as HashMap<*, *>
        val imaSdkSettings = ImaSdkFactory.getInstance().createImaSdkSettings()

        if (adsLoaderSettings["ppid"] is String) {
            imaSdkSettings.ppid = adsLoaderSettings["ppid"] as String
        }
        imaSdkSettings.language = adsLoaderSettings["language"] as String
        imaSdkSettings.isDebugMode = adsLoaderSettings["enable_debug_mode"] as Boolean

        val uri = payload["uri"] as String;
        val imaPlayerSettings = ImaPlayerSettings(uri, payload["ima_tag"] as? String?)

        imaPlayerSettings.isMixed = payload["is_mixed"] as Boolean;
        imaPlayerSettings.autoPlay = payload["auto_play"] as Boolean;
        imaPlayerSettings.initialVolume = payload["initial_volume"] as Double;
        imaPlayerSettings.showPlaybackControls = payload["show_playback_controls"] as Boolean

        var headers = HashMap<String, String>()
        if (payload["headers"] is HashMap<*, *>) {
            headers = payload["headers"] as HashMap<String, String>
        }

        if (uri.startsWith("asset://")) {
            val flutterLoader = FlutterInjector.instance().flutterLoader()
            val assetKey = flutterLoader.getLookupKeyForAsset(uri.substring(8))
            imaPlayerSettings.uri = Uri.parse("asset:///${assetKey}")
        }

        return ImaPlayerView(
            context,
            id,
            messenger,
            imaSdkSettings,
            imaPlayerSettings,
            headers
        )
    }
}

File: ios/Classes/ImaPlayerViewFactory.swift
//
//  ImaPlayerViewFactory.swift
//  ima_player
//
//  Created by Ömer Güven on 19.07.2023.
//

import GoogleInteractiveMediaAds
import Foundation
import Flutter
import UIKit


class ImaPlayerViewFactory: NSObject, FlutterPlatformViewFactory {
    private var registrar: FlutterPluginRegistrar

    init(registrar: FlutterPluginRegistrar) {
        self.registrar = registrar;
        super.init()
    }
    
    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
          return FlutterStandardMessageCodec.sharedInstance()
    }

    func create(withFrame frame: CGRect,  viewIdentifier viewId: Int64, arguments args: Any? ) -> FlutterPlatformView {
        var methodChannel = FlutterMethodChannel(name: "gece.dev/imaplayer/\(viewId)", binaryMessenger: registrar.messenger())
        var eventChannel = FlutterEventChannel(name: "gece.dev/imaplayer/\(viewId)/events", binaryMessenger: registrar.messenger())
        
        let payload = args as! Dictionary<String, Any>
        
        let adsLoaderSettings = payload["ads_loader_settings"] as! Dictionary<String, Any>
        let imaSdkSettings  = IMASettings()
        
        
        if adsLoaderSettings["ppid"] is String {
            imaSdkSettings.ppid = (adsLoaderSettings["ppid"] as! String)
        }
    
        imaSdkSettings.language = adsLoaderSettings["language"] as! String
        imaSdkSettings.enableBackgroundPlayback = true
        
        var imaPlayerSettings = ImaPlayerSettings()
        let uri = payload["uri"] as! String;
        
        imaPlayerSettings.tag = payload["ima_tag"] as? String
        imaPlayerSettings.uri = URL(string: uri)
        imaPlayerSettings.autoPlay = payload["auto_play"] as! Bool
        imaPlayerSettings.initialVolume = payload["initial_volume"] as! Double
        imaPlayerSettings.isMixed = payload["is_mixed"] as! Bool
        imaPlayerSettings.showPlaybackControls = payload["show_playback_controls"] as! Bool
        
        var headers = Dictionary<String, String>()
        
        if (payload["headers"] is Dictionary<String, Any>) {
            headers = payload["headers"] as! Dictionary<String, String>
        }
        
        if uri.starts(with: "asset://") {
            let assetKey = registrar.lookupKey(forAsset: String(uri.dropFirst(8)))
            let assetPath = Bundle.main.path(forResource: assetKey, ofType: nil)!
            imaPlayerSettings.uri = URL(string: assetPath, relativeTo: Bundle.main.bundleURL)
        }
        
        return ImaPlayerView(
            frame: frame,
            viewIdentifier: viewId,
            arguments: args as! Dictionary<String, Any>,
            binaryMessenger: registrar.messenger(),
            imaSdkSettings: imaSdkSettings,
            imaPlayerSettings: imaPlayerSettings,
            headers: headers,
            methodChannel: methodChannel,
            eventChannel: eventChannel
        )
    }
}
